feat: add ability to view compose logs

cdrage opened the pull request:
feat: add ability to view compose logs

### What does this PR do?

When clicking on the group of containers, you can now view the logs of
the entire group of compose containers.

### Screenshot/screencast of this PR

<!-- Please include a screenshot or a screencast explaining what is doing this PR -->

![Screenshot 2023-07-11 at 12 48 47 PM](https://github.com/containers/podman-desktop/assets/6422176/743a4ffc-e291-4697-8ac5-8052cc921946)


### What issues does this PR fix or reference?

<!-- Please include any related issue from Podman Desktop repository (or from another issue tracker).
-->

Closes https://github.com/containers/podman-desktop/issues/3103

### How to test this PR?

1. Deploy a docker-compose example
(https://raw.githubusercontent.com/kubernetes/kompose/main/examples/docker-compose.yaml)
2. Go to the container view in PD
3. Press on the container group
4. View the logs

<!-- Please explain steps to reproduce -->

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

benoitf commented on packages/renderer/src/lib/compose/ComposeDetailsLogs.svelte:
```diff
@@ -0,0 +1,186 @@
+// An array of readable ANSI escape sequence colours against a black terminal background
+// these are the most "readable" colours against a black background
+// No colours like grey, normal blue (cyan instead) or red, since they don't appear very well.
+const ansi256Colors = [
```
It looks like pods logs and composeLogs page could share common code instead of copy paste between these two ?

cdrage replied:
you're right, I can move the colourization to another file. 

The other functions I wouldn't be able to move since it ties in quite tightly with iTerm, etc.

I'll refactor this! thanks for pointing it out.

cdrage reviewed (commented):
Added the "editor" shared code to a new file.

I also fixed the spelling of color to colour :)

jeffmaury commented on packages/renderer/src/lib/compose/ComposeDetails.spec.ts:
```diff
@@ -0,0 +1,42 @@
+  render(ComposeDetails, { composeName: 'foobar', engineId: 'engine' });
+  // Click on the logs href
+  const logsHref = screen.getByRole('link', { name: 'Logs' });
+  fireEvent.click(logsHref);
```
Maybe good to add an assertion to check that the content of the log tab is present

cdrage replied:
Done! Added an assertion to check that it's loaded in the test (although it's empty screen).

I'm unsure if we should add a more inclusive test for loading the containers, I was thinking of coordinating with @odockal and getting more of an integration test added since we do not mock xterm

jeffmaury commented on packages/renderer/src/lib/compose/ComposeDetailsLogs.svelte:
```diff
@@ -0,0 +1,165 @@
+const colourizedContainerName = new Map<string, string>();
+
+// Callback for logs which will output the logs to the terminal
+function callback(name: string, data: string) {
```
Maybe rename to have a more meaningful name lilke `onLogMessage` or `onLogData` ?

cdrage replied:
copied-and-pasted from the podman logs section, I'd rather keep the same name.

jeffmaury commented on packages/renderer/src/lib/compose/ComposeDetailsLogs.svelte:
```diff
@@ -0,0 +1,165 @@
+    colourizedContainerName.set(container.name, colourizedANSIContainerName(container.name, colour));
+  });
+
+  console.log('colors:', colourizedContainerName);
```
Should be removed ?

lstocchi commented on packages/renderer/src/lib/compose/ComposeDetails.svelte:
```diff
@@ -0,0 +1,93 @@
+
+{#if compose}
+  <DetailsPage title="{composeName}" subtitle="">
+    <StatusIcon slot="icon" icon="{ComposeIcon}" status="{compose.status}" />
```
It would be good to use `svelte:fragment` here to define slots https://github.com/containers/podman-desktop/pull/3213

deboer-tim replied:
When there's only one object getting put in the slot IMHO it's shorter and simpler to tag it like this vs creating a wrapper.

lstocchi commented on packages/renderer/src/App.svelte:
```diff
@@ -128,7 +129,10 @@ window.events?.receive('display-troubleshooting', () => {
             resourceId="{decodeURI(meta.params.resourceId)}"
             engineId="{decodeURI(meta.params.engineId)}" />
         </Route>
-        <Route path="/pods/:kind/:name/:engineId/*" breadcrumb="Pod Details" let:meta navigationHint="details">
```
Did you forget to add the navigationHint in the Pod details route?

deboer-tim replied:
Good catch, the new page should have it as well. You can check the breadcrumb (especially after selecting an action and cancelling it) to confirm correct behaviour.

lstocchi reviewed (commented):
Everything works fine, just a couple of comments below and a graphical issue.

The buttons in the group container miss the background
This is what i see
![image](https://github.com/containers/podman-desktop/assets/49404737/802aa632-7334-4505-bfd0-3e421289d342)

but in a simple container page, it's like this
![image](https://github.com/containers/podman-desktop/assets/49404737/c445bdf4-f878-46d5-b5ab-bec7babeb9ce)

cdrage commented:
> Everything works fine, just a couple of comments below and a graphical issue.
> 
> The buttons in the group container miss the background This is what i see ![image](https://user-images.githubusercontent.com/49404737/254854923-802aa632-7334-4505-bfd0-3e421289d342.png)
> 
> but in a simple container page, it's like this ![image](https://user-images.githubusercontent.com/49404737/254855132-c445bdf4-f878-46d5-b5ab-bec7babeb9ce.png)

Ah you're right. I had forgotten to add `detailed=true` to when I imported ComposeActions.

I've updated the PR so that the buttons now display correctly:

![Screenshot 2023-07-20 at 10 12 02 AM](https://github.com/containers/podman-desktop/assets/6422176/c1ddbe15-f735-451b-8f81-193973bfadc1)

lstocchi reviewed (approved):
LGTM!! Thanks!!

cdrage commented:
Going to merge this in as it's been reviewed by @benoitf , changed and updated, then re-reviewed twice by @lstocchi 

Thanks for being patient with all of this!
